driver/src/price_estimation/kraken/api.rs

```diff
@@ -46,35 +50,50 @@ impl KrakenHttpApi {
 }
 
 impl KrakenApi for KrakenHttpApi {
-    fn assets(&self) -> Result<HashMap<String, Asset>> {
-        self.client
-            .get_json::<_, KrakenResult<_>>(format!("{}/Assets", self.base_url), HttpLabel::Kraken)
-            .context("failed to parse assets JSON")?
-            .into_result()
+    fn assets<'a>(&'a self) -> BoxFuture<'a, Result<HashMap<String, Asset>>> {
+        async move {
+            self.client
+                .get_json_async::<_, KrakenResult<_>>(
+                    format!("{}/Assets", self.base_url),
+                    HttpLabel::Kraken,
+                )
+                .await
+                .context("failed to parse assets JSON")?
+                .into_result()
+        }
+        .boxed()
     }
 
-    fn asset_pairs(&self) -> Result<HashMap<String, AssetPair>> {
-        self.client
-            .get_json::<_, KrakenResult<_>>(
-                format!("{}/AssetPairs", self.base_url),
-                HttpLabel::Kraken,
-            )
-            .context("failed to parse asset pairs JSON")?
-            .into_result()
+    fn asset_pairs<'a>(&'a self) -> BoxFuture<'a, Result<HashMap<String, AssetPair>>> {
+        async move {
+            self.client
+                .get_json_async::<_, KrakenResult<_>>(
+                    format!("{}/AssetPairs", self.base_url),
+                    HttpLabel::Kraken,
+                )
+                .await
+                .context("failed to parse asset pairs JSON")?
+                .into_result()
+        }
+        .boxed()
     }
 
-    fn ticker(&self, pairs: &[&str]) -> Result<HashMap<String, TickerInfo>> {
+    fn ticker<'a, 'b>(
+        &'a self,
+        pairs: &'b [&'b str],
+    ) -> BoxFuture<'a, Result<HashMap<String, TickerInfo>>> {
         if pairs.is_empty() {
-            return Ok(HashMap::new());
+            return async { Ok(HashMap::new()) }.boxed();
         }
-
-        self.client
-            .get_json::<_, KrakenResult<_>>(
-                format!("{}/Ticker?pair={}", self.base_url, pairs.join(",")),
-                HttpLabel::Kraken,
-            )
-            .context("failed to parse ticker JSON")?
-            .into_result()
+        let url = format!("{}/Ticker?pair={}", self.base_url, pairs.join(","));
+        async move {
+            self.client
+                .get_json_async::<_, KrakenResult<_>>(url, HttpLabel::Kraken)
+                .await
+                .context("failed to parse ticker JSON")?
+                .into_result()
+        }
+        .boxed()
     }
```

Comment:
The reason for having two async blocks in one function is that this allows us to make the resulting Future's lifetime only depend on `self`.

Reply:
Can't you you just make `pairs: &'a [&'a str]`?

Reply:
Alternatively, you could make `url: Option<String>` and then have it be `None` when pairs are empty. That allows you to move the `if` statement into the `async` block and not need two of them.

Reply:
> Can't you you just make `pairs: &'a [&'a str]`? I feel like the two `async` blocks are a little awkward.

That needlessly restricts the lifetimes of `pairs` to be the same as self which means it gets tied to the resulting Future.

I changed it to use your second suggestion.